The following is the error result when attempting to compile with the last push.


nbramhall3@ece-rschsrv.ece.gatech.edu>make xrun
looking in file Include/calculator.include
cd /nethome/nbramhall3/Digital-Design-Onboarding/sim/behav/WORKSPACE && \
xrun -64bit -sv -linedebug -access +rwc -timescale 1ns/10ps +define+SIM=1 +define+INST_FILE=\"\" +testname= +incdir+sym_links -coverage all -licqueue -covoverwrite -f sym_links/sim_no_path.include \
     -logfile simulation.log
TOOL:   xrun(64)        25.03-s003: Started on Sep 11, 2025 at 21:53:08 EDT
xrun(64): 25.03-s003: (c) Copyright 1995-2025 Cadence Design Systems, Inc.
xrun: *W,BADPRF: The -linedebug option may have an adverse performance impact.
Recompiling... reason: file './sym_links/controller.sv' is newer than expected.
        expected: Thu Sep 11 21:52:26 2025
        actual:   Thu Sep 11 21:52:59 2025
file: sym_links/controller.sv
        module worklib.controller:sv
                errors: 0, warnings: 0
file: sym_links/top_lvl.sv
xmvlog: *W,SPDUSD: Include directory sym_links given but not used.
        Total errors/warnings found outside modules and primitives:
                errors: 0, warnings: 1
                Caching library 'worklib' ....... Done
        Elaborating the design hierarchy:
        Top level design units:
                calculator_pkg
                $unit_0x3f332ce4
                tb_calculator

        Extracting FSMs for coverage:
                worklib.controller
                worklib.sky130_sram_2kbyte_1rw1r_32x512_8
                worklib.full_adder
                worklib.adder32
                worklib.result_buffer
                worklib.top_lvl
                worklib.tb_calculator
        Total FSMs extracted = 0
        .wmask0 (wmask0), 
                      |
xmelab: *W,CUVMPW (./sym_links/top_lvl.sv,82|22): port sizes differ in port connection(1/4) for the instance(tb_calculator.DUT) .
        .din0   (w_data),
                      |
xmelab: *W,CUVMPW (./sym_links/top_lvl.sv,84|22): port sizes differ in port connection(64/32) for the instance(tb_calculator.DUT) .
        .dout1  (r_data) 
                      |
xmelab: *W,CUVMPW (./sym_links/top_lvl.sv,89|22): port sizes differ in port connection(64/32) for the instance(tb_calculator.DUT) .
        .wmask0 (wmask0), 
                      |
xmelab: *W,CUVMPW (./sym_links/top_lvl.sv,98|22): port sizes differ in port connection(1/4) for the instance(tb_calculator.DUT) .
        .din0   (w_data),
                      |
xmelab: *W,CUVMPW (./sym_links/top_lvl.sv,100|22): port sizes differ in port connection(64/32) for the instance(tb_calculator.DUT) .
        .dout1  (r_data) 
                      |
xmelab: *W,CUVMPW (./sym_links/top_lvl.sv,105|22): port sizes differ in port connection(64/32) for the instance(tb_calculator.DUT) .
  .w_data(w_data),
        |
xmelab: *E,MULDRN (./sym_links/top_lvl.sv,53|8): Variable 'tb_calculator.DUT.w_data' has multiple conflicting drivers and the indicated output port connection in scope 'tb_calculator.DUT' is contributing to this conflict set.
      .buffer_o(w_data)  //output to sram
              |
xmelab: *E,MULDRN (./sym_links/top_lvl.sv,121|14): Variable 'tb_calculator.DUT.w_data' has multiple conflicting drivers and the indicated output port connection in scope 'tb_calculator.DUT' is contributing to this conflict set.
        .dout1  (r_data) 
             |
xmelab: *E,MULDRN (./sym_links/top_lvl.sv,89|13): Variable 'tb_calculator.DUT.r_data' has multiple conflicting drivers and the indicated output port connection in scope 'tb_calculator.DUT' is contributing to this conflict set.
        .dout1  (r_data) 
             |
xmelab: *E,MULDRN (./sym_links/top_lvl.sv,105|13): Variable 'tb_calculator.DUT.r_data' has multiple conflicting drivers and the indicated output port connection in scope 'tb_calculator.DUT' is contributing to this conflict set.
        assign write = 1;
                     |
xmelab: *E,ICDPAV (./sym_links/controller.sv,100|14): Illegal combination of driver and procedural assignment to variable write detected (procedural assignment found in always block at line 117 in file ./sym_links/controller.sv).
        assign read = 1;
                    |
xmelab: *E,ICDPAV (./sym_links/controller.sv,99|13): Illegal combination of driver and procedural assignment to variable read detected (procedural assignment found in always block at line 112 in file ./sym_links/controller.sv).
        assign state = next;
                     |
xmelab: *E,ICDPAV (./sym_links/controller.sv,102|14): Illegal combination of driver and procedural assignment to variable state detected (procedural assignment found in always block at line 48 in file ./sym_links/controller.sv).
xrun: *E,ELBERR: Error (*E) or soft error (*SE) occurred during elaboration (status 1), exiting.
TOOL:   xrun(64)        25.03-s003: Exiting on Sep 11, 2025 at 21:53:09 EDT  (total: 00:00:01)
make: *** [Makefile:74: xrun] Error 1
nbramhall3@ece-rschsrv.ece.gatech.edu>

--- a/verilog/adder32.sv
+++ b/verilog/adder32.sv
@@ -15,11 +15,11 @@ module adder32 (
 
     genvar i;
     generate
-        for (i = 0; i <  32; i = i + 1) begin
+        for (i = 0; i <  DATA_W; i = i + 1) begin : gen_adders_block
 	        full_adder fa_inst (
                 .a      (a_i[i]),
                 .b      (b_i[i]),
-                .cin    (i == 0 ? 1'b0 : c[i-1]),
+                .cin    (c[i]),
                 .s      (sum_o[i]),
                 .cout   (c[i+1])
             );

--- a/verilog/adder32.sv
+++ b/verilog/adder32.sv
@@ -1,7 +1,8 @@
 /*
 * Module describing a 32-bit ripple carry adder, with no carry output or input
 */
-module adder32 import calculator_pkg::*; (
+import calculator_pkg::*;
+module adder32 (
     input logic [DATA_W - 1 : 0] a_i,
     input logic [DATA_W - 1 : 0] b_i,
     output logic [DATA_W - 1 : 0] sum_o

--- a/verilog/controller.sv
+++ b/verilog/controller.sv
@@ -19,15 +19,17 @@ module controller (
   	// Writing Control
     output logic write,							//flag, when 0 = write.
     output logic [ADDR_W-1:0] w_addr,			//address to write data to
-    output logic [MEM_WORD_SIZE-1:0] w_data,	//the data thats being written
+    output logic [MEM_WORD_SIZE-1:32] w_data_a, //the data thats being written to sram a
+    output logic [31:0] w_data_b,				//the data thats being written to sram b
 
 	//Reading Controls
-    input  logic [MEM_WORD_SIZE-1:0] r_data,	//data thats being read
+    input  logic [MEM_WORD_SIZE-1:32] r_data_a,	//data thats being read from sram a
+	input  logic [31:0] 			  r_data_b,	//data thats being read from sram b
 	output logic read,							//flag, when 0 = read
     output logic [ADDR_W-1:0] r_addr,			//address thats being read from
 	
 	input  logic [MEM_WORD_SIZE-1:0] buff_result,// the result thats being stored in the buffer and will be written to SRAM
-    output logic              buffer_control,	 // Buffer Control (1 = upper, 0, = lower)
+    output logic              		 buffer_control,	 // Buffer Control (1 = upper, 0, = lower)
 
   	output logic [DATA_W-1:0]       op_a,		//input into adder
     output logic [DATA_W-1:0]       op_b		//input into adder
@@ -68,26 +70,36 @@ module controller (
 	//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 	//curr_radder register
 	always_ff @(posedge clk_i) begin
-		if (state == S_IDLE) begin
-			curr_raddr <= read_start_addr;
-			
-		end else if (state == S_READ) begin
-			curr_raddr <= curr_raddr + 1; 	//moves address forward
+		if (rst_i) begin 
+			curr_raddr <= 'b0;
+		end else begin
+			if (state == S_IDLE) begin
+				curr_raddr <= read_start_addr;
+			end else if (state == S_READ) begin
+				curr_raddr <= curr_raddr + 1; 	//moves address forward
+			end
 		end
 	end 
 
 	//curr_wadder register
 	always_ff @(posedge clk_i) begin
-		if(state == S_IDLE) begin	
-			curr_waddr <= write_start_addr;
+		if (rst_i) begin 
+			curr_waddr <= 'b0;
+		end else begin
+			if(state == S_IDLE) begin	
+				curr_waddr <= write_start_addr;
+			end
 		end
 	end
 
 	//buffer control register
 	always_ff @(posedge clk_i) begin
-		buffer_control <= 0;
+		if (rst_i) begin 
+			buffer_control <= 'b0;
+		end else begin
 			if (state == S_ADD) begin
 				buffer_control <= ~buffer_control;
+			end
 		end
 	end
 
@@ -96,15 +108,16 @@ module controller (
 	//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 	assign r_addr = curr_raddr;
 	assign w_addr = curr_waddr;
-	assign read = 1;
-	assign write = 1;
-	assign w_data = buff_result;
-	assign state = next;
+	assign w_data_a = buff_result [MEM_WORD_SIZE - 1: 0];
+	assign w_data_b = buff_result [31:0];
 
 	always_comb begin
 		if (state == S_ADD)begin
-				op_a = r_data[DATA_W-1:32];						//assgines the data from the address to the adder input a
-				op_b = r_data[31:0]; 					//assgines the data from the address to the adder input b
+			op_a = r_data_a;						//assgines the data from the address to the adder input a
+			op_b = r_data_b; 					//assgines the data from the address to the adder input b
+		end else begin
+			op_a = 0;
+			op_b = 0;
 		end
 	end
 	

--- a/verilog/result_buffer.sv
+++ b/verilog/result_buffer.sv
@@ -24,11 +24,13 @@ module result_buffer  (
             internal_buffer <= 0;
         end else begin
             if (loc_sel) begin
-                buffer_o[MEM_WORD_SIZE - 1 : 32] <= result_i;
+                internal_buffer[MEM_WORD_SIZE - 1 : 32] <= {result_i, internal_buffer[31:0]};
             end else begin
-                buffer_o[31 : 0] <= result_i;
+                internal_buffer[31 : 0] <= {internal_buffer[MEM_WORD_SIZE-1:32], result_i};
             end
             //Place result_i into buffer based on loc_sel
         end
     end
+	 
+	 assign buffer_o = internal_buffer;
 endmodule

--- a/verilog/result_buffer.sv
+++ b/verilog/result_buffer.sv
@@ -24,9 +24,9 @@ module result_buffer  (
             internal_buffer <= 0;
         end else begin
             if (loc_sel) begin
-                internal_buffer[MEM_WORD_SIZE - 1 : 32] <= {result_i, internal_buffer[31:0]};
+                internal_buffer <= {result_i, internal_buffer[31:0]};
             end else begin
-                internal_buffer[31 : 0] <= {internal_buffer[MEM_WORD_SIZE-1:32], result_i};
+                internal_buffer <= {internal_buffer[MEM_WORD_SIZE-1:32], result_i};
             end
             //Place result_i into buffer based on loc_sel
         end

--- a/verilog/top_lvl.sv
+++ b/verilog/top_lvl.sv
@@ -22,39 +22,43 @@ module top_lvl (
     logic [ADDR_W-1:0] w_addr;
     logic [ADDR_W-1:0] r_addr;
 
-    logic [MEM_WORD_SIZE-1:0] w_data; 
-    logic [MEM_WORD_SIZE-1:0] r_data;
+    logic [MEM_WORD_SIZE-1:32] w_data_a; 
+    logic [31:0] w_data_b;
+    logic [MEM_WORD_SIZE-1:32] r_data_a;
+    logic [31:0] r_data_b;
 
     logic [DATA_W-1:0] op_a; 
     logic [DATA_W-1:0] op_b;
     logic [DATA_W-1:0] sum_o;
 
     logic [MEM_WORD_SIZE-1:0] buff_result; 
-
+    logic buffer_control;
     //SRAM wire
-    logic wmask0 = 4'hF;
+    logic [3:0] wmask0 = 4'hF;
 
     //TODO: Finish instantiation of your controller module
 	controller u_ctrl (
-  //inputs
-  .clk_i(clk),
-  .rst_i(rst),
-  .read_start_addr(read_start_addr),
-  .read_end_addr(read_end_addr),
-  .write_start_addr(write_start_addr),
-  .write_end_addr(write_end_addr),
-  .r_data(r_data),
-  .buff_result(buff_result),
+    //inputs
+    .clk_i            (clk),
+    .rst_i            (rst),
+    .read_start_addr  (read_start_addr),
+    .read_end_addr    (read_end_addr),
+    .write_start_addr (write_start_addr),
+    .write_end_addr    (write_end_addr),
+    .r_data_a         (r_data_a),
+    .r_data_b         (r_data_b),
+    .buff_result      (buff_result),
 
-  //outputs
-  .write(write),
-  .read(read),
-  .w_addr(w_addr),
-  .w_data(w_data),
-  .r_addr(r_addr),
-  .buffer_control(buffer_control),  //1 = upper, 0= lower
-  .op_a(op_a),
-  .op_b(op_b)
+    //outputs
+    .write          (write),
+    .read           (read),
+    .w_addr         (w_addr),
+    .w_data_a       (w_data_a),  
+    .w_data_b       (w_data_b),
+    .r_addr         (r_addr),
+    .buffer_control (buffer_control),  //1 = upper, 0= lower
+    .op_a           (op_a),
+    .op_b           (op_b)
   );
 
     //TODO: Look at the sky130_sram_2kbyte_1rw1r_32x512_8 module and instantiate it using variables defined above.
@@ -80,13 +84,13 @@ module top_lvl (
         .csb0   ((read | write)),
         .web0   (write), 
         .wmask0 (wmask0), 
-        .addr0  (w_addr), //maybe r_addr 
-        .din0   (w_data),
+        .addr0  (w_addr), 
+        .din0   (w_data_a),
         .dout0  (), //leave blank
         .clk1   (clk), 
         .csb1   (read), 
         .addr1  (r_addr), 
-        .dout1  (r_data) 
+        .dout1  (r_data_a) 
     );
 
   
@@ -97,27 +101,27 @@ module top_lvl (
         .web0   (write), 
         .wmask0 (wmask0), 
         .addr0  (w_addr), //maybe r_addr 
-        .din0   (w_data),
+        .din0   (w_data_b),
         .dout0  (), //leave blank
         .clk1   (clk), 
         .csb1   (read), 
         .addr1  (r_addr), 
-        .dout1  (r_data) 
+        .dout1  (r_data_b) 
     );
   	
   	//TODO: Finish instantiation of your adder module
     adder32 u_adder (
-      .a_i(op_a), //input from srama
-      .b_i(op_b), //input from sramb
-      .sum_o(sum_o) //output to buffer
+      .a_i      (op_a), //input from srama
+      .b_i      (op_b), //input from sramb
+      .sum_o    (sum_o) //output to buffer
     );
  
     //TODO: Finish instantiation of your result buffer
     result_buffer u_resbuf (
-      .clk_i(clk),
-      .rst_i(rst),
-      .result_i(sum_o), //input from adder
-      .loc_sel(buffer_control),  //input
-      .buffer_o(w_data)  //output to sram
+      .clk_i    (clk),
+      .rst_i    (rst),
+      .result_i (sum_o), //input from adder
+      .loc_sel  (buffer_control),  //input
+      .buffer_o (buff_result)  //output to sram
     );
 endmodule

